src/miyoo_sdk.c

```diff
@@ -646,94 +631,354 @@ void gp_Reset(void)
+		uint16_t l1[320];
+		uint16_t l2[320];
+		uint16_t* l3;
+
+		for (y = 0; y < 16; y++)
 		{
-			pSource+=32;
-			for (x = 64; x != 0; x--)
-			{
-				uint16_t _1 = pSource[0];
-				pTarget[0] = _1;
-				uint16_t _2 = pSource[1];
-				pTarget[1] = Weight1_3( _1, _2);
-				uint16_t _3 = pSource[2];
-				pTarget[2] = Weight1_1( _2, _3);
-				uint16_t _4 = pSource[3];
-				pTarget[3] = Weight3_1( _3, _4);
-				pTarget[4] = _4;
-				pTarget+=5;
-				pSource+=4;		
-			}
-			pSource+=32;
+			x = 64;
+			// Direct copy
+			pSource += 32;
+			ScaleLineSlow(pTarget, pSource, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(pTarget, pSource, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(pTarget, pSource, x);
+			l3 = pTarget;
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			// Interpolate 1/3
+			ScaleLineSlow(l2, pSource, x);
+			InterpolateLines_1_3(pTarget, l3, l2, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(l1, pSource, x);
+			InterpolateLines_1_3(pTarget, l2, l1, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(l2, pSource, x);
+			InterpolateLines_1_3(pTarget, l1, l2, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			// Interpolate 1/1
+			ScaleLineSlow(l1, pSource, x);
+			InterpolateLines_1_1(pTarget, l2, l1, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(l2, pSource, x);
+			InterpolateLines_1_1(pTarget, l1, l2, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(l1, pSource, x);
+			InterpolateLines_1_1(pTarget, l2, l1, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			// Interpolate 3/1
+			ScaleLineSlow(l2, pSource, x);
+			InterpolateLines_3_1(pTarget, l1, l2, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(l1, pSource, x);
+			InterpolateLines_3_1(pTarget, l2, l1, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			l3 = pTarget + 320;
+			ScaleLineSlow(l3, pSource, x);
+			InterpolateLines_3_1(pTarget, l1, l3, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			// Direct copy
+			pTarget += 320;
+
+			ScaleLineSlow(pTarget, pSource, x);
+			pSource += 256 + 64;
+			pTarget += 320;
+
+			ScaleLineSlow(pTarget, pSource, x);
+			pSource += 256 + 32;
+			pTarget += 320;
 		}
 	}
-	else // 224
+	else // H == 239
 	{
-		pSource += 2560;
-		unsigned short pos = 2;
-		for (y = 240; y != 0; y--)
+		for (y = H; y != 0; y--)
 		{
-			pSource+=32;
-			for (x = 64; x != 0; x--)
-			{
-				uint16_t _1 = pSource[0];
-				pTarget[0] = _1;
-				uint16_t _2 = pSource[1];
-				pTarget[1] = Weight1_3( _1, _2);
-				uint16_t _3 = pSource[2];
-				pTarget[2] = Weight1_1( _2, _3);
-				uint16_t _4 = pSource[3];
-				pTarget[3] = Weight3_1( _3, _4);
-				pTarget[4] = _4;
-				pTarget+=5;
-				pSource+=4;		
-			}
-			pSource+=32;
-			pos--;
+			x = 64;
+			pSource += 32;
+			ScaleLineSlow(pTarget, pSource, x);
+			pSource += 256 + 32;
+			pTarget += 320;
+		}
+	}
+}
 
-			if (pos == 0)
```

Comment:
do u happen to know why there's position check for 224 source height in fullscreen mode? I noticed you've discarded that part from your scalers.

Reply:
This corresponds to the old vertical scaling algorithm, which duplicates a line every 14 lines (calculating it twice). In the new algorithm, the vertical pattern is hard-coded, so it's no longer needed.